src/cargo/util/config/mod.rs

```diff
@@ -764,6 +765,26 @@ impl Config {
         }
     }
 
+    /// Get the value of environment variable `key` through the `Config` snapshot.
+    ///
+    /// This can be used similarly to `std::env::var`.
+    pub fn get_env(&self, key: impl AsRef<str>) -> CargoResult<String> {
+        match self.env.get(key.as_ref()) {
+            Some(s) => Ok(s.clone()),
+            None => bail!(
+                "{} could not be found in the environment snapshot",
+                key.as_ref()
+            ),
+        }
+    }
+
+    /// Get the value of environment variable `key` through the `Config` snapshot.
+    ///
+    /// This can be used similarly to `std::env::var_os`.
+    pub fn get_env_os(&self, key: impl AsRef<str>) -> Option<String> {
+        self.env.get(key.as_ref()).cloned()
+    }
```

Comment:
This might be a little confusing since it is not returning an `OsString`, which is what the `_os` suffix implies.

The lack of `OsString` support also seems like a regression in the cases where `var_os` is used. Cargo has fairly poor support for non-UTF-8 paths and environment, but I think it would be good to try to avoid making it worse.

How hard would it be to store `OsString` in the map instead of `String`?

Reply:
Oops yes good point! I'll investigate what else needs to change if `Config.env` is a `HashMap<OsString, OsString>` instead.

Reply:
Ok, this turned out to be not too bad to implement (I think). Please see the most recent commit. I mainly had to add/edit a few utility methods on `Config` to handle invalid UTF-8 entries in `self.env`.